File: advent-of-code/2015/03/02/solution.zig
const std = @import("std");

fn countHousesWithRobo(directions: []const u8) !u32 {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var santa_x: i32 = 0;
    var santa_y: i32 = 0;
    var robo_x: i32 = 0;
    var robo_y: i32 = 0;
    var visited = std.AutoHashMap([2]i32, void).init(allocator);
    defer visited.deinit();

    try visited.put([2]i32{ 0, 0 }, {});

    for (directions, 0..) |c, i| {
        if (i % 2 == 0) { // Santa's turn
            switch (c) {
                '^' => santa_y += 1,
                'v' => santa_y -= 1,
                '>' => santa_x += 1,
                '<' => santa_x -= 1,
                else => {},
            }
            try visited.put([2]i32{ santa_x, santa_y }, {});
        } else { // Robo-Santa's turn
            switch (c) {
                '^' => robo_y += 1,
                'v' => robo_y -= 1,
                '>' => robo_x += 1,
                '<' => robo_x -= 1,
                else => {},
            }
            try visited.put([2]i32{ robo_x, robo_y }, {});
        }
    }

    return visited.count();
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();

    // Test cases with assertions
    std.debug.assert(try countHousesWithRobo("^v") == 3);
    std.debug.assert(try countHousesWithRobo("^>v<") == 3);
    std.debug.assert(try countHousesWithRobo("^v^v^v^v^v") == 11);

    const input = try std.fs.cwd().readFileAlloc(gpa.allocator(), "../input.txt", 1024 * 1024);
    defer gpa.allocator().free(input);
    const directions = std.mem.trim(u8, input, &std.ascii.whitespace);

    const result = try countHousesWithRobo(directions);
    std.debug.print("Zig: {d}\n", .{result});
}

File: advent-of-code/2015/03/02/solution.rs
use std::collections::HashSet;
use std::fs;

fn count_houses_with_robo(directions: &str) -> usize {
    let mut santa_x = 0;
    let mut santa_y = 0;
    let mut robo_x = 0;
    let mut robo_y = 0;
    let mut visited = HashSet::new();
    visited.insert((0, 0));
    
    for (i, c) in directions.chars().enumerate() {
        if i % 2 == 0 { // Santa's turn
            match c {
                '^' => santa_y += 1,
                'v' => santa_y -= 1,
                '>' => santa_x += 1,
                '<' => santa_x -= 1,
                _ => {}
            }
            visited.insert((santa_x, santa_y));
        } else { // Robo-Santa's turn
            match c {
                '^' => robo_y += 1,
                'v' => robo_y -= 1,
                '>' => robo_x += 1,
                '<' => robo_x -= 1,
                _ => {}
            }
            visited.insert((robo_x, robo_y));
        }
    }
    
    visited.len()
}

fn main() {
    // Test cases with assertions
    assert_eq!(count_houses_with_robo("^v"), 3);
    assert_eq!(count_houses_with_robo("^>v<"), 3);
    assert_eq!(count_houses_with_robo("^v^v^v^v^v"), 11);

    // Read input and calculate result
    let directions = fs::read_to_string("../input.txt")
        .expect("Failed to read input file")
        .trim()
        .to_string();
    
    let result = count_houses_with_robo(&directions);
    println!("Rust: {}", result);
}
